### PR TITLE
min() max() now throw NPE on null values

### LINKED ISSUE #1482: NPE in max() and min()
Using min or max on a list containing a null value will either produce a NPE or Option.none:

```
    @Test
    public void nullPointerExceptionInMinAndMax() {
        Integer one = 1;
        Integer zero = 0;
        Integer nullInteger = null;

        try {
            List<Integer> nullSecond = List.of(zero, nullInteger, one);
            nullSecond.min();
            nullSecond.max();
        } catch (NullPointerException e) {
            System.out.println("Boooo. Gives NPE but should give min == zero && max == one. Due to o1.compareTo(o2) in max and min");
        }

        List<Integer> nullAsHead = List.of(nullInteger, zero, one);
        System.out.println("nullAsHead.min() = " + nullAsHead.min());
        System.out.println("nullAsHead.max() = " + nullAsHead.max());
        System.out.println("Boo gives None in both situations. should give min == zero && max = one. Due to the fact: !(list.head() instanceof Comparable))");

        try {
            List<Integer> nullAtEnd = List.of(zero, one, nullInteger);
            nullAtEnd.min();
            nullAtEnd.max();
        } catch (NullPointerException e) {
            System.out.println("Boooo. Gives NPE but should give min == zero && max == one. Due to o1.compareTo(o2) in max and min");
        }
    }
```

### PR DESCRIPTION
**Pull request by danieldietrich**
Fixes #1482

### CODE REVIEW

File: javaslang/src/test/java/javaslang/TestComparators.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ *    /  /    \/ \  / \/    \ /  /\__\/  //    \/  \  //  /\__\   JΛVΛSLΛNG
+ *  _/  /  /\  \  \/  /  /\  \\__\\  \  //  /\  \ /\\/ \ /__\ \   Copyright 2014-2016 Javaslang, http://javaslang.io
+ * /___/\_/  \_/\____/\_/  \_/\__\/__/\__\_/  \_//  \__/\_____/   Licensed under the Apache License, Version 2.0
+ */
+package javaslang;
+
+import java.io.Serializable;
+import java.util.Comparator;
+
+public final class TestComparators {
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/TestComparators.java]: \[Checkstyle\] ERROR: Utility classes should not have a public or default constructor\.

File: javaslang/src/test/java/javaslang/collection/AbstractSortedSetTest.java [hunk 2]
```diff
@@ -9,17 +9,22 @@
...
+import java.util.Comparator;
+
+import static javaslang.TestComparators.toStringComparator;
 
 public abstract class AbstractSortedSetTest extends AbstractSetTest {
 
     @Override
     abstract protected <T> SortedSet<T> of(T element);
 
+    abstract protected <T> SortedSet<T> of(Comparator<? super T> comparator, T element);
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/AbstractSortedSetTest.java]: \[Checkstyle\] ERROR: 'protected' modifier out of order with the JLS suggestions\.

File: javaslang/src/test/java/javaslang/collection/BitSetTest.java [hunk 3]
```diff
@@ -125,6 +126,12 @@ protected boolean emptyShouldBeSingleton() {
         return this.<T> bsBuilder().of(element);
     }
 
+    @Override
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/BitSetTest.java]: \[Checkstyle\] ERROR: Fields and methods should be before inner classes\.

File: javaslang/src/test/java/javaslang/collection/TreeMultimapTest.java [hunk 4]
```diff
@@ -1,6 +1,7 @@
 package javaslang.collection;
 
 import javaslang.Tuple2;
+import javaslang.Value;
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/TreeMultimapTest.java]: \[Checkstyle\] ERROR: Unused import \- javaslang\.Value\.

File: javaslang/src/test/java/javaslang/collection/TreeSetTest.java [hunk 5]
```diff
@@ -39,40 +38,29 @@ protected boolean emptyShouldBeSingleton() {
...
-            }
-        }
-        if (allLongs) {
-            return TreeSet.ofAll(toLongComparator(), Iterator.of(elements));
-        } else if (allNumbers) {
-            return TreeSet.ofAll(toDoubleComparator(), Iterator.of(elements));
-        } else {
-            return TreeSet.ofAll(toStringComparator(), Iterator.of(elements));
-        }
+        return TreeSet.<T> of(naturalComparator(), elements);
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/TreeSetTest.java]: \[Checkstyle\] ERROR: '>' is followed by whitespace\.

Comment (danieldietrich) [inline javaslang/src/test/java/javaslang/collection/TreeSetTest.java]: @pivovarit do you have a hint how I can tweak Sputnik's default rules to accept this whitespace?

File: javaslang/src/test/java/javaslang/collection/AbstractSortedSetTest.java (hunk 2 above)
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/AbstractSortedSetTest.java]: \[Checkstyle\] ERROR: 'protected' modifier out of order with the JLS suggestions\.

File: javaslang/src/test/java/javaslang/collection/BitSetTest.java [hunk 6]
```diff
@@ -125,6 +99,12 @@ protected boolean emptyShouldBeSingleton() {
...
+    protected <T> BitSet<T> of(Comparator<? super T> comparator, T element) {
+        // comparator is not used
+        return this.<T> bsBuilder().of(element);
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/BitSetTest.java]: \[Checkstyle\] ERROR: '>' is followed by whitespace\.

File: javaslang/src/test/java/javaslang/collection/BitSetTest.java [hunk 7]
```diff
@@ -524,4 +504,33 @@ public void shouldConvertToSortedSetWithoutComparatorOnComparable() {
...
+
+    // -- classes
+
+    private static class Mapper<T> implements Serializable {
+
+        private static final long serialVersionUID = 1L;
+
+        private final java.util.Map<Integer, T> fromIntMap = new java.util.HashMap<>();
+        private final java.util.Map<T, Integer> toIntMap = new java.util.HashMap<>();
+        private int nextValue = 0;
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/BitSetTest.java]: \[Checkstyle\] ERROR: Variable 'nextValue' explicitly initialized to '0' \(default value for its type\)\.

File: javaslang/src/test/java/javaslang/collection/TreeSetTest.java (hunk 5 above)
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/TreeSetTest.java]: \[Checkstyle\] ERROR: '>' is followed by whitespace\.

File: javaslang/src/test/java/javaslang/collection/euler/PrimeNumbers.java [hunk 8]
```diff
@@ -11,18 +11,18 @@
 import javaslang.collection.Stream;
 import javaslang.collection.TreeSet;
 
-public final class PrimeNumbers {
+final class PrimeNumbers {
 
-    private static Set<Integer> PRIMES_2_000_000 = Sieve.fillSieve.apply(2_000_000, TreeSet.empty());
+    private static Set<Integer> PRIMES_2_000_000 = Sieve.fillSieve(2_000_000, TreeSet.empty());
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/euler/PrimeNumbers.java]: \[Checkstyle\] ERROR: Name 'PRIMES\_2\_000\_000' must match pattern '^\[a\-z\]\[a\-zA\-Z0\-9\]\*$'\.

File: javaslang/src/test/java/javaslang/collection/euler/Sieve.java [hunk 9]
```diff
@@ -10,15 +10,18 @@
 import javaslang.collection.*;
 import javaslang.control.Option;
 
-public class Sieve {
+class Sieve {
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/euler/Sieve.java]: \[Checkstyle\] ERROR: Class Sieve should be declared as final\.

File: javaslang/src/test/java/javaslang/collection/euler/Sieve.java [hunk 10]
```diff
@@ -10,15 +10,18 @@
...
 
-    final static List<Function2<Integer, Integer, Option<Integer>>> rules = List.of(
+    private Sieve() {
+    }
+
+    private final static List<Function2<Integer, Integer, Option<Integer>>> rules = List.of(
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/euler/Sieve.java]: \[Checkstyle\] ERROR: 'static' modifier out of order with the JLS suggestions\.

Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/euler/Sieve.java]: \[Checkstyle\] ERROR: Name 'rules' must match pattern '^\(log|\[A\-Z\]\[A\-Z0\-9\]\*\(\_\[A\-Z0\-9\]\+\)\*\)$'\.

File: javaslang/src/test/java/javaslang/collection/euler/Sieve.java [hunk 11]
```diff
@@ -10,15 +10,18 @@
...
             (x, y) -> Option.of((4 * x * x) + (y * y)).filter(n -> n % 12 == 1 || n % 12 == 5),
             (x, y) -> Option.of((3 * x * x) + (y * y)).filter(n -> n % 12 == 7),
             (x, y) -> Option.of((3 * x * x) - (y * y)).filter(n -> x > y && n % 12 == 11)
     );
 
-    final static List<Function3<Set<Integer>, Integer, Integer, Set<Integer>>> steps = List.of(
+    private final static List<Function3<Set<Integer>, Integer, Integer, Set<Integer>>> steps = List.of(
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/euler/Sieve.java]: \[Checkstyle\] ERROR: 'static' modifier out of order with the JLS suggestions\.

Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/euler/Sieve.java]: \[Checkstyle\] ERROR: Name 'steps' must match pattern '^\(log|\[A\-Z\]\[A\-Z0\-9\]\*\(\_\[A\-Z0\-9\]\+\)\*\)$'\.

File: javaslang/src/test/java/javaslang/collection/euler/Utils.java [hunk 12]
```diff
@@ -13,40 +15,40 @@
...
-public final class Utils {
+final class Utils {
 
     private Utils() {
     }
 
-    public static Stream<BigInteger> fibonacci() {
+    static final Function1<Integer, BigInteger> memoizedFactorial = Function1.of(Utils::factorial).memoized();
+
+    static final Function1<Long, Boolean> memoizedIsPrime = Function1.of(Utils::isPrime).memoized();
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/euler/Utils.java]: \[Checkstyle\] ERROR: Name 'memoizedIsPrime' must match pattern '^\(log|\[A\-Z\]\[A\-Z0\-9\]\*\(\_\[A\-Z0\-9\]\+\)\*\)$'\.

**Comment by danieldietrich** [comment]
@pivovarit is there a way to perform the Sputnik checks locally in order to get faster turn-around cycles?

File: javaslang/src/test/java/javaslang/collection/AbstractSortedSetTest.java (hunk 2 above)
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/AbstractSortedSetTest.java]: \[Checkstyle\] ERROR: 'protected' modifier out of order with the JLS suggestions\.

File: javaslang/src/test/java/javaslang/collection/BitSetTest.java (hunk 6 above)
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/BitSetTest.java]: \[Checkstyle\] ERROR: '>' is followed by whitespace\.

File: javaslang/src/test/java/javaslang/collection/BitSetTest.java (hunk 7 above)
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/BitSetTest.java]: \[Checkstyle\] ERROR: Variable 'nextValue' explicitly initialized to '0' \(default value for its type\)\.

File: javaslang/src/test/java/javaslang/collection/TreeSetTest.java (hunk 5 above)
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/TreeSetTest.java]: \[Checkstyle\] ERROR: '>' is followed by whitespace\.

File: javaslang/src/test/java/javaslang/collection/euler/Sieve.java [hunk 13]
```diff
@@ -7,21 +7,26 @@
...
 import javaslang.control.Option;
 
-public class Sieve {
+final class Sieve {
 
-    final static List<Function2<Integer, Integer, Option<Integer>>> rules = List.of(
+    private Sieve() {
+    }
+
+    private final static List<Function2<Integer, Integer, Option<Integer>>> RULES = List.of(
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/euler/Sieve.java]: \[Checkstyle\] ERROR: 'static' modifier out of order with the JLS suggestions\.

File: javaslang/src/test/java/javaslang/collection/euler/Sieve.java [hunk 14]
```diff
@@ -7,21 +7,26 @@
...
             (x, y) -> Option.of((4 * x * x) + (y * y)).filter(n -> n % 12 == 1 || n % 12 == 5),
             (x, y) -> Option.of((3 * x * x) + (y * y)).filter(n -> n % 12 == 7),
             (x, y) -> Option.of((3 * x * x) - (y * y)).filter(n -> x > y && n % 12 == 11)
     );
 
-    final static List<Function3<Set<Integer>, Integer, Integer, Set<Integer>>> steps = List.of(
+    private final static List<Function3<Set<Integer>, Integer, Integer, Set<Integer>>> STEPS = List.of(
```
Comment (sputnikci) [inline javaslang/src/test/java/javaslang/collection/euler/Sieve.java]: \[Checkstyle\] ERROR: 'static' modifier out of order with the JLS suggestions\.

**Comment by danieldietrich** [comment]
The remaining Sputnik remarks are ok for now - need to tweak the rules a bit.

File: javaslang/src/test/java/javaslang/collection/AbstractTraversableTest.java [hunk 15]
```diff
@@ -968,6 +968,16 @@ public void shouldComputeMaxOfBigDecimal() {
         assertThat(of(BigDecimal.ZERO, BigDecimal.ONE).max()).isEqualTo(Option.some(BigDecimal.ONE));
     }
 
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowNPEWhenMaxOfNullAndInt() {
+        System.out.println(of(null, 1).max());
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/AbstractTraversableTest.java]: I think you can get rid of the `System.out.println`

File: javaslang/src/test/java/javaslang/collection/AbstractTraversableTest.java [hunk 16]
```diff
@@ -968,6 +968,16 @@ public void shouldComputeMaxOfBigDecimal() {
...
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowNPEWhenMaxOfIntAndNull() {
+        System.out.println(of(1, null).max());
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/AbstractTraversableTest.java]: same

File: javaslang/src/test/java/javaslang/collection/AbstractTraversableTest.java [hunk 17]
```diff
@@ -1084,6 +1094,16 @@ public void shouldComputeMinOfBigDecimal() {
...
     }
 
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowNPEWhenMinOfNullAndInt() {
+        of(null, 1).min();
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowNPEWhenMinOfIntAndNull() {
+        System.out.println(of(1, null).min());
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/AbstractTraversableTest.java]: same

File: javaslang/src/test/java/javaslang/collection/BitSetTest.java (hunk 6 above)
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/BitSetTest.java]: Should be Synchronized with the Idea styles :)

File: javaslang/src/test/java/javaslang/collection/PriorityQueueTest.java [hunk 18]
```diff
@@ -150,11 +146,17 @@ public void shouldPreserveSingletonInstanceOnDeserialization() {
...
+        // makes no sense because sorted sets contain ordered elements
+    }
+
     // -- static narrow
 
+    @SuppressWarnings("unchecked")
     @Test
-    public void shouldNarrowQueue() {
-        final PriorityQueue<Double> doubles = of(1.0d);
+    public final void shouldNarrowPriorityQueue() {
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/PriorityQueueTest.java]: `final`?

File: javaslang/src/test/java/javaslang/collection/euler/Utils.java [hunk 19]
```diff
@@ -13,40 +15,40 @@
...
                 .filter(d -> number % d == 0)
                 .flatMap(d -> Stream.of(d, number / d))
                 .distinct();
     }
 
-    public static Stream<Long> divisors(long l) {
-        return factors(l)
-                .filter((d) -> d < l);
+    static Stream<Long> divisors(long l) {
+        return factors(l).filter((d) -> d < l);
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/collection/euler/Utils.java]: `(d) ->` can be `d ->`

File: javaslang/src/test/java/javaslang/concurrent/Concurrent.java [hunk 20]
```diff
@@ -13,16 +13,19 @@
 
 import static org.assertj.core.api.Assertions.fail;
 
-interface Concurrent {
+final class Concurrent {
 
-    Random RND = new Random();
+    private static final Random RND = new Random();
```
Comment (paplorinc) [inline javaslang/src/test/java/javaslang/concurrent/Concurrent.java]: could also be: `RND = ThreadLocalRandom.current();`

**Comment by pivovarit** [comment]
@danieldietrich running Sputnik locally is not possible but you can definitely run those tools that Sputnik uses (those can be included as a part of the build). Before doing this, it's good to create a local config for those tools (which will help with the whitespace problem mentioned earlier)

File: javaslang/src/test/java/javaslang/collection/TreeSetTest.java (hunk 5 above)
Comment (pivovarit) [inline javaslang/src/test/java/javaslang/collection/TreeSetTest.java]: There is a way of overriding the default config by adding some files locally. There it should be possible to tweak it.

@pjagielski can you help?

Comment (pjagielski) [inline javaslang/src/test/java/javaslang/collection/TreeSetTest.java]: Yes, under project' settings, there is a "I have my own config" checkbox which allows you to select a file from repo to use as checkstyle/findbugs/pmd config: 

image

image

Comment (danieldietrich) [inline javaslang/src/test/java/javaslang/collection/TreeSetTest.java]: Thank you, that helps!

**Pull request merged**